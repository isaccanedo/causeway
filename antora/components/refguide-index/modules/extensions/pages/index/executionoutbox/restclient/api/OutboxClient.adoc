= OutboxClient
:Notice: Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at. http://www.apache.org/licenses/LICENSE-2.0 . Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR  CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

== API

[source,java]
.OutboxClient.java
----
class OutboxClient {
  OutboxClient(String restfulBaseUrl, String username, String password)
  OutboxClient(String restfulBaseUrl, String tenantId, String clientId, String clientSecret)
  OutboxClient(RestfulClientConfig restfulClientConfig)
  OutboxClient(RestfulClientConfig restfulClientConfig, OutboxClientConfig outboxClientConfig)
  OutboxClient(RestfulClientConfig restfulClientConfig, AuthorizationHeaderFactory authorizationHeaderFactory)
  OutboxClient(RestfulClientConfig restfulClientConfig, AuthorizationHeaderFactory authorizationHeaderFactory, OutboxClientConfig outboxClientConfig)
  OutboxClient withConnectTimeoutInSecs(int connectTimeoutInSecs)
  OutboxClient withReadTimeoutInSecs(int readTimeoutInSecs)
  List<InteractionDto> pending()
  void delete(String interactionId, int sequence)
  void deleteMany(List<InteractionDto> interactionDtos)
}
----

