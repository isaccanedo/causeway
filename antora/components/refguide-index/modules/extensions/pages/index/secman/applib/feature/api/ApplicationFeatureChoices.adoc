= ApplicationFeatureChoices
:Notice: Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at. http://www.apache.org/licenses/LICENSE-2.0 . Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR  CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

Supports mixins to add and filter permissions of _org.apache.causeway.extensions.secman.applib.role.dom.ApplicationRole_ and _org.apache.causeway.extensions.secman.applib.user.dom.ApplicationUser_ .

== API

[source,java]
.ApplicationFeatureChoices.java
----
class ApplicationFeatureChoices {
  public static final String DESCRIBED_AS;
  Collection<ApplicationFeatureChoices.AppFeat> autoCompleteFeature(String search)
}
----

