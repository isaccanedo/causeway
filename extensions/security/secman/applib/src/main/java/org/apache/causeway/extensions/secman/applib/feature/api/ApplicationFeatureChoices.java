/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.extensions.secman.applib.feature.api;

import java.util.Collection;
import java.util.Map;
import java.util.Objects;
import java.util.TreeSet;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Named;

import org.springframework.stereotype.Service;

import org.apache.causeway.applib.ViewModel;
import org.apache.causeway.applib.annotation.DomainObject;
import org.apache.causeway.applib.annotation.MinLength;
import org.apache.causeway.applib.annotation.Nature;
import org.apache.causeway.applib.annotation.ObjectSupport;
import org.apache.causeway.applib.annotation.Property;
import org.apache.causeway.applib.services.appfeat.ApplicationFeatureId;
import org.apache.causeway.applib.services.appfeat.ApplicationFeatureRepository;
import org.apache.causeway.commons.internal.functions._Predicates;
import org.apache.causeway.extensions.secman.applib.CausewayModuleExtSecmanApplib;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;

/**
 * Supports mixins to add and filter permissions of
 * {@link org.apache.causeway.extensions.secman.applib.role.dom.ApplicationRole}
 * and {@link org.apache.causeway.extensions.secman.applib.user.dom.ApplicationUser}.
 *
 * @since 2.0 {@index}
 */
@Service
@RequiredArgsConstructor(onConstructor_ = {@Inject})
public class ApplicationFeatureChoices {

    final ApplicationFeatureRepository featureRepository;

    public static final String DESCRIBED_AS =
            "To refine the search by feature-sort (namespace, type, member), "
            + "use one of "
            + "n: t: m:";

    public Collection<ApplicationFeatureChoices.AppFeat> autoCompleteFeature(
            final @MinLength(3) String search) {

        final Predicate<ApplicationFeatureId> searchRefine;
        final String searchTerm;

        if(search.startsWith("n:")) {
            searchRefine = ApplicationFeatureChoices::isNamespace;
            searchTerm = search.substring(2).trim();
        } else if(search.startsWith("t:")) {
            searchRefine = ApplicationFeatureChoices::isType;
            searchTerm = search.substring(2).trim();
        } else if(search.startsWith("m:")) {
            searchRefine = ApplicationFeatureChoices::isMember;
            searchTerm = search.substring(2).trim();
        } else {
            searchRefine = _Predicates.alwaysTrue();
            searchTerm = search.trim();
        }

        var idsByName = featureRepository.getFeatureIdentifiersByName();

        return idsByName.entrySet().stream()
        .filter(entry->searchRefine.test(entry.getValue()))
        .filter(entry -> entry.getKey().contains(searchTerm))
        .map(Map.Entry::getValue)
        .map(ApplicationFeatureChoices.AppFeat::new)
        .collect(Collectors.toCollection(TreeSet::new));
    }

    private static boolean isNamespace(final ApplicationFeatureId featureId) {
        return featureId.getSort().isNamespace();
    }

    private static boolean isType(final ApplicationFeatureId featureId) {
        return featureId.getSort().isType();
    }

    private static boolean isMember(final ApplicationFeatureId featureId) {
        return featureId.getSort().isMember();
    }

    // -- FEATURE VIEW MODEL WRAPPING A VALUE TYPE

    /**
     * Viewmodel wrapper around value type {@link ApplicationFeatureId}. Introduced,
     * because at the time of writing,
     * autoComplete/choices do not support value types.
     */
    @Named(AppFeat.LOGICAL_TYPE_NAME)
    @DomainObject(
            nature = Nature.VIEW_MODEL)
    @AllArgsConstructor @NoArgsConstructor @EqualsAndHashCode
    public static class AppFeat
    implements
        Comparable<AppFeat>,
        ViewModel {

        public static final String LOGICAL_TYPE_NAME = CausewayModuleExtSecmanApplib.NAMESPACE + ".AppFeat";

        @Property
        @Getter
        private ApplicationFeatureId featureId;

        @ObjectSupport public String title() {
            return toString();
        }

        @Override
        public int compareTo(final AppFeat o) {
            var thisId = this.getFeatureId();
            var otherId = o!=null ? o.getFeatureId() : null;
            if(Objects.equals(thisId, otherId)) {
                return 0;
            }
            if(thisId==null) {
                return -1;
            }
            if(otherId==null) {
                return 1;
            }
            return this.getFeatureId().compareTo(o.getFeatureId());
        }

        @Override
        public String toString() {
            return featureId!=null
                    ? featureId.getSort().name() + ": " + featureId.getFullyQualifiedName()
                    : "<no id>";
        }

        // -- VIEWMODEL CONTRACT

        @Inject
        public AppFeat(final String memento) {
            this(ApplicationFeatureId.parseEncoded(memento)); // fail by intention if memento is '<no id>'
        }

        @Override
        public String viewModelMemento() {
            return featureId!=null
                    ? featureId.asEncodedString()
                    : "<no id>";
        }

    }

}
