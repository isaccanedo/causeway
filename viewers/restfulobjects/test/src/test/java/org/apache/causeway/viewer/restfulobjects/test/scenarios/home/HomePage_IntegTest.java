/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.restfulobjects.test.scenarios.home;

import javax.ws.rs.client.Invocation;
import javax.ws.rs.core.Response;

import org.apache.causeway.viewer.restfulobjects.test.scenarios.Abstract_IntegTest;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class HomePage_IntegTest extends Abstract_IntegTest {

    @Test
    @UseReporter(DiffReporter.class)
    public void homePage() {

        // given
        Invocation.Builder request = restfulClient.request("/");

        // when
        var response = request.get();

        // then
        assertThat(response)
                .extracting(Response::getStatus)
                .isEqualTo(Response.Status.OK.getStatusCode());

        var entity = response.readEntity(String.class);

        Approvals.verify(entity, jsonOptions());

    }
}
