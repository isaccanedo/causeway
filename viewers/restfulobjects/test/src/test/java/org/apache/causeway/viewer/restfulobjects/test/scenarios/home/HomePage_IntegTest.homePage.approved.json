{
  "links" : [ {
    "rel" : "self",
    "href" : "http://0.0.0.0:NNN/restful/",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/homepage\""
  }, {
    "rel" : "urn:org.restfulobjects:rels/user",
    "href" : "http://0.0.0.0:NNN/restful/user",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/user\""
  }, {
    "rel" : "urn:org.apache.causeway.restfulobjects:rels/menuBars",
    "href" : "http://0.0.0.0:NNN/restful/menuBars",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/layout-menubars\""
  }, {
    "rel" : "urn:org.restfulobjects:rels/services",
    "href" : "http://0.0.0.0:NNN/restful/services",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/list\""
  }, {
    "rel" : "urn:org.restfulobjects:rels/version",
    "href" : "http://0.0.0.0:NNN/restful/version",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/version\""
  }, {
    "rel" : "urn:org.restfulobjects:rels/domain-types",
    "href" : "http://0.0.0.0:NNN/restful/domain-types",
    "method" : "GET",
    "type" : "application/json;profile=\"urn:org.restfulobjects:repr-types/type-list\""
  } ],
  "extensions" : { }
}